File: src/interop/module_config.ts
import { invoke } from "@tauri-apps/api/core";
import { OsisBook } from "./bible";
import { HtmlText } from "./html_text";

export interface ExternalModuleData {
    aliases: Record<string, string>;
    assets: Record<string, string>;
}

export type ReadingsFormat = |{
    type: "daily";
    count: number;
} |{
    type: "weekly";
    count: number;
} |{
    type: "yearly";
    count: number;
    leap_year: "skip" | number;
};

export interface BibleConfig {
  type: "bible_config";
  name: string;
  id: string;
  authors?: string[];
  language?: string;
  description?: HtmlText;
  data_source?: string;
  pub_year?: number;
  license?: string;
  books: Record<OsisBook, string>;
  book_abbreviations: Record<OsisBook, string>;
  book_aliases: Record<string, OsisBook>;
  external: ExternalModuleData;
}

export interface NotebookConfig {
  type: "notebook_config";
  name: string;
  id: string;
  authors?: string[];
  language?: string;
  description?: HtmlText;
  data_source?: string;
  pub_year?: number;
  bible?: string;
  external: ExternalModuleData;
}

export interface StrongsDefConfig {
  type: "strongs_def_config";
  name: string;
  id: string;
  authors?: string[];
  language?: string;
  description?: HtmlText;
  data_source?: string;
  pub_year?: number;
  license?: string;
  external: ExternalModuleData;
}

export interface StrongsLinksConfig {
  type: "strongs_links_config";
  name: string;
  id: string;
  short_name?: string;
  bible: string;
  authors?: string[];
  language?: string;
  description?: HtmlText;
  data_source?: string;
  pub_year?: number;
  license?: string;
  external: ExternalModuleData;
}

export interface CommentaryConfig {
  type: "commentary_config";
  name: string;
  id: string;
  short_name?: string;
  bible?: string;
  authors?: string[];
  language?: string;
  description?: HtmlText;
  data_source?: string;
  pub_year?: number;
  license?: string;
  external: ExternalModuleData;
}

export interface DictionaryConfig {
    type: "dictionary_config";
    name: string;
    id: string;
    short_name?: string;
    authors?: string[];
    language?: string;
    description?: HtmlText;
    data_source?: string;
    pub_year?: number;
    license?: string;
    external: ExternalModuleData;
}

export interface ReadingsConfig {
    type: "readings_config";
    name: string;
    id: string;
    short_name?: string;
    authors?: string[];
    description?: HtmlText;
    data_source?: string;
    pub_year?: number;
    license?: string;
    language?: string;
    format: ReadingsFormat;
    external: ExternalModuleData;
}

export interface XRefConfig {
    type: "x_ref_config";
    name: string;
    id: string;
    short_name?: string;
    authors?: string[];
    language?: string;
    description?: HtmlText;
    data_source?: string;
    pub_year?: number;
    license?: string;
    bible?: string;
    external: ExternalModuleData;
}

export type ModuleConfig =
  | BibleConfig
  | NotebookConfig
  | StrongsDefConfig
  | StrongsLinksConfig
  | CommentaryConfig
  | DictionaryConfig
  | ReadingsConfig
  | XRefConfig;


export async function fetch_backend_module_configs(): Promise<ModuleConfig[]>
{
    return await invoke<string>("run_bible_command", {
        command: {
            type: "fetch_module_configs"
        }
    }).then(v => {
        return JSON.parse(v);
    })
}

File: src-tauri/src/repr/module_config.rs
use std::collections::HashMap;

use biblio_json::{core::{OsisBook, lang::Language}, modules::{ExternalModuleData, Module, ModuleId, bible::BibleModule, readings::ReadingsFormat}};
use serde::{Deserialize, Serialize};

use crate::repr::HtmlTextJson;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case", tag = "type")]
pub enum ModuleConfigJson
{
    BibleConfig
    {
        name: String,
        id: ModuleId,
        authors: Option<Vec<String>>,
        language: Option<Language>,
        description: Option<HtmlTextJson>,
        data_source: Option<String>,
        pub_year: Option<u32>,
        license: Option<String>,
        books: HashMap<OsisBook, String>,
        book_abbreviations: HashMap<OsisBook, String>,
        book_aliases: HashMap<String, OsisBook>,
        external: ExternalModuleData,
    },
    NotebookConfig
    {
        name: String,
        id: ModuleId,
        authors: Option<Vec<String>>,
        language: Option<Language>,
        description: Option<HtmlTextJson>,
        data_source: Option<String>,
        pub_year: Option<u32>,
        bible: Option<ModuleId>,
        external: ExternalModuleData,
    },
    StrongsDefConfig
    {
        name: String,
        id: ModuleId,
        authors: Option<Vec<String>>,
        language: Option<Language>,
        description: Option<HtmlTextJson>,
        data_source: Option<String>,
        pub_year: Option<u32>,
        license: Option<String>,
        external: ExternalModuleData,
    },
    StrongsLinksConfig
    {
        name: String,
        id: ModuleId,
        short_name: Option<String>,
        bible: ModuleId,
        authors: Option<Vec<String>>,
        language: Option<Language>,
        description: Option<HtmlTextJson>,
        data_source: Option<String>,
        pub_year: Option<u32>,
        license: Option<String>,
        external: ExternalModuleData,
    },
    CommentaryConfig
    {
        name: String,
        id: ModuleId,
        short_name: Option<String>,
        bible: Option<ModuleId>,
        authors: Option<Vec<String>>,
        language: Option<Language>,
        description: Option<HtmlTextJson>,
        data_source: Option<String>,
        pub_year: Option<u32>,
        license: Option<String>,
        external: ExternalModuleData,
    },
    DictionaryConfig
    {
        name: String,
        id: ModuleId,
        short_name: Option<String>,
        authors: Option<Vec<String>>,
        language: Option<Language>,
        description: Option<HtmlTextJson>,
        data_source: Option<String>,
        pub_year: Option<u32>,
        license: Option<String>,
        external: ExternalModuleData,
    },
    ReadingsConfig
    {
        name: String,
        id: ModuleId,
        short_name: Option<String>,
        authors: Option<Vec<String>>,
        description: Option<HtmlTextJson>,
        data_source: Option<String>,
        pub_year: Option<u32>,
        license: Option<String>,
        language: Option<Language>,
        format: ReadingsFormat,
        external: ExternalModuleData,
    },
    XRefConfig
    {
        name: String,
        id: ModuleId,
        short_name: Option<String>,
        authors: Option<Vec<String>>,
        language: Option<Language>,
        description: Option<HtmlTextJson>,
        data_source: Option<String>,
        pub_year: Option<u32>,
        license: Option<String>,
        bible: Option<ModuleId>,
        external: ExternalModuleData,
    },
}

impl ModuleConfigJson
{
    pub fn new(module: &Module) -> Self
    {
        match module 
        {
            Module::Bible(bible) => {
                let config = &bible.config;
                Self::BibleConfig { 
                    name: config.name.clone(), 
                    id: config.id.clone(), 
                    authors: config.authors.clone(), 
                    language: config.language.clone(), 
                    description: config.description.as_ref().map(|d| HtmlTextJson::from_html(&d, &config.external)), 
                    data_source: config.data_source.clone(), 
                    pub_year: config.pub_year, 
                    license: config.license.clone(), 
                    books: config.books.clone(), 
                    book_abbreviations: config.book_abbreviations.clone(), 
                    book_aliases: config.book_aliases.clone(), 
                    external: config.external.clone() 
                }
            },
            Module::Dictionary(m) => {
                let config = &m.config;
                Self::DictionaryConfig { 
                    name: config.name.clone(), 
                    id: config.id.clone(), 
                    short_name: config.short_name.clone(), 
                    authors: config.authors.clone(), 
                    language: config.language.clone(), 
                    description: config.description.as_ref().map(|d| HtmlTextJson::from_html(&d, &config.external)), 
                    data_source: config.data_source.clone(), 
                    pub_year: config.pub_year, 
                    license: config.license.clone(), 
                    external: config.external.clone()
                }
            },
            Module::XRef(m) => {
                let config = &m.config;
                Self::XRefConfig {
                    name: config.name.clone(),
                    id: config.id.clone(),
                    short_name: config.short_name.clone(),
                    authors: config.authors.clone(),
                    language: config.language.clone(),
                    description: config.description.as_ref().map(|d| HtmlTextJson::from_html(&d, &config.external)),
                    data_source: config.data_source.clone(),
                    pub_year: config.pub_year,
                    license: config.license.clone(),
                    bible: config.bible.clone(),
                    external: config.external.clone(),
                }
            },
            Module::StrongsDefs(m) => {
                let config = &m.config;
                Self::StrongsDefConfig {
                    name: config.name.clone(),
                    id: config.id.clone(),
                    authors: config.authors.clone(),
                    language: config.language.clone(),
                    description: config.description.as_ref().map(|d| HtmlTextJson::from_html(&d, &config.external)),
                    data_source: config.data_source.clone(),
                    pub_year: config.pub_year,
                    license: config.license.clone(),
                    external: config.external.clone(),
                }
            },
            Module::StrongsLinks(m) => {
                let config = &m.config;
                Self::StrongsLinksConfig {
                    name: config.name.clone(),
                    id: config.id.clone(),
                    short_name: config.short_name.clone(),
                    bible: config.bible.clone(),
                    authors: config.authors.clone(),
                    language: config.language.clone(),
                    description: config.description.as_ref().map(|d| HtmlTextJson::from_html(&d, &config.external)),
                    data_source: config.data_source.clone(),
                    pub_year: config.pub_year,
                    license: config.license.clone(),
                    external: config.external.clone(),
                }
            },
            Module::Commentary(m) => {
                let config = &m.config;
                Self::CommentaryConfig {
                    name: config.name.clone(),
                    id: config.id.clone(),
                    short_name: config.short_name.clone(),
                    bible: config.bible.clone(),
                    authors: config.authors.clone(),
                    language: config.language.clone(),
                    description: config.description.as_ref().map(|d| HtmlTextJson::from_html(&d, &config.external)),
                    data_source: config.data_source.clone(),
                    pub_year: config.pub_year,
                    license: config.license.clone(),
                    external: config.external.clone(),
                }
            },
            Module::Notebook(m) => {
                let config = &m.config;
                Self::NotebookConfig {
                    name: config.name.clone(),
                    id: config.id.clone(),
                    authors: config.authors.clone(),
                    language: config.language.clone(),
                    description: config.description.as_ref().map(|d| HtmlTextJson::from_html(&d, &config.external)),
                    data_source: config.data_source.clone(),
                    pub_year: config.pub_year,
                    bible: config.bible.clone(),
                    external: config.external.clone(),
                }
            },
            Module::Readings(m) => {
                let config = &m.config;
                Self::ReadingsConfig {
                    name: config.name.clone(),
                    id: config.id.clone(),
                    short_name: config.short_name.clone(),
                    authors: config.authors.clone(),
                    description: config.description.as_ref().map(|d| HtmlTextJson::from_html(&d, &config.external)),
                    data_source: config.data_source.clone(),
                    pub_year: config.pub_year,
                    license: config.license.clone(),
                    language: config.language.clone(),
                    format: config.format.clone(),
                    external: config.external.clone(),
                }
            },
        }
    }
}